usaspending_api/references/v2/views/cfda.py

```diff
@@ -0,0 +1,48 @@
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from requests import post
+from time import sleep
+from django.conf import settings
+
+ALL_CFDAS = None
+
+
+class CFDAViewSet(APIView):
+    """
+    Return an agency name and active fy.
+    """
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/v2/references/agency/id.md"
+
+    def get(self, request, cfda=None):
+        """
+        Return the view's queryset.
+        """
+        self._populate_cfdas_if_needed()
+
+        if cfda:
+            results = [result for result in ALL_CFDAS if result["cfda"] == cfda]
+        else:
+            results = ALL_CFDAS
+
+        response = {"results": results}
+
+        return Response(response)
+
+    def _populate_cfdas_if_needed(self):
+        global ALL_CFDAS
+        if not ALL_CFDAS:
+            response = self._request_from_grants_api()
+
+            #  grants API is brittle in practice, so if we don't get results retry at a polite rate
+            while not response:
```

Comment:
As for the infinite loop, yes, it would be possible if the grants API went down indefinitely, although our daily redeploy means that no request would hog resources permanently. Either way, it would be easy enough to add a counter to just "give up" after a minute since we wouldn't be returning anything to the user after that point anyway.

It is true that if Grants API gives a different response structure than what was expected, this code would throw an exception. In that case, however, I would argue that the correct response would be to fail, and 500 would be the most appropriate response code, since a service used on the server side of the architecture failed.

Reply:
> Additionally, it seems there is an expectation an exception will be raised in _request_from_grants_api() from .json()["cfdas"] since that is making hard assumptions about the response.

The context of that comment was for breaking out of the infinite loop. There was no explicit checking or error handling to help end the infinite loop.

In regards to your comment, it would be a good idea to add helpful logging to know if the issue was unreachable server, unexpected API response, authentication/missing api, or whatever likely situation you expect.